### PR TITLE
[Messenger] Add support for the DelayStamp in InMemoryTransport

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | n/a
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Review by welcoMattic — APPROVED** [review]

File: src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php [hunk 1]
```diff
@@ -55,7 +57,15 @@ public function __construct(SerializerInterface $serializer = null)
 
     public function get(): iterable
     {
-        return array_values($this->decode($this->queue));
+        $envelopes = [];
+        $now = new \DateTimeImmutable();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php]: let's inject a ClockInterface?

Comment (fabpot) [reply src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php]: Done

File: src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php [hunk 2]
```diff
@@ -46,16 +48,25 @@ class InMemoryTransport implements TransportInterface, ResetInterface
...
     private int $nextId = 1;
-    private ?SerializerInterface $serializer;
+    private array $availableAt = [];
 
-    public function __construct(SerializerInterface $serializer = null)
-    {
-        $this->serializer = $serializer;
+    public function __construct(
+        private ?SerializerInterface $serializer = null,
+        private ?ClockInterface $clock = null,
```
Comment (OskarStark) [inline src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php:55]: Do we need this as require dev dep? Or is it already available?

File: src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php [hunk 3]
```diff
@@ -46,16 +48,25 @@ class InMemoryTransport implements TransportInterface, ResetInterface
...
+    ) {
     }
 
     public function get(): iterable
     {
-        return array_values($this->decode($this->queue));
+        $envelopes = [];
+        $now = $this->clock ? $this->clock->now() : new \DateTimeImmutable();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php]: `$now = $this->clock?->now() : new \DateTimeImmutable();`

File: src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php [hunk 4]
```diff
@@ -11,8 +11,10 @@
 
 namespace Symfony\Component\Messenger\Transport\InMemory;
 
+use Symfony\Component\Clock\ClockInterface;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php]: let's use Psr\Clock\ClockInterface

File: src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php [hunk 5]
```diff
@@ -88,6 +99,11 @@ public function send(Envelope $envelope): Envelope
         $this->sent[] = $encodedEnvelope;
         $this->queue[$id] = $encodedEnvelope;
 
+        /** @var DelayStamp|null $delayStamp */
+        if ($delayStamp = $envelope->last(DelayStamp::class)) {
+            $this->availableAt[$id] = (new \DateTimeImmutable())->modify(sprintf('+%d seconds', $delayStamp->getDelay() / 1000));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php]: `$now = $this->clock?->now() : new \DateTimeImmutable();` here also

File: src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php]: Yep

File: src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php (hunk 3 above)
Comment (fabpot) [reply src/Symfony/Component/Messenger/Transport/InMemory/InMemoryTransport.php]: ok

**Review by nicolas-grekas — APPROVED** [review]

**Pull request merged**